### ISSUE TITLE
Feature Request: Toggle between frame and keyframe onion skin

### ISSUE THREAD

**Issue by Gilraiser** [synfig/synfig#1817]
**Synfig version & platform**:

Development version
Archlinux 64

**Description**:

I think it is very important to have the frames and also the keyframes as a reference in the work area to create movements because this helps us to achieve greater precision. I suggest an icon or icons where we can enable onion skin for frames and keyframes. Or if you want to have both activated so that they can be seen simultaneously.

Attached is a recreation of how it could be.

Thank you very much in advance for your attention.

Thanks for the excellent work.

![1Osk](https://user-images.githubusercontent.com/45057390/97767602-28424300-1aeb-11eb-8a3b-e73b46476a51.gif)
![2OSk](https://user-images.githubusercontent.com/45057390/97767795-84f22d80-1aec-11eb-87f1-992a49bacd64.gif)
![3OSk](https://user-images.githubusercontent.com/45057390/97767920-71939200-1aed-11eb-8063-b700f65a7510.gif)

**Comment by DhairyaBahl** [comment]
@ice0 I would like to work on this issue. Kindly assign it to me. Thanks :)

**Comment by morevnaproject** [comment]
@DhairyaBahl Assigned to you. ^__^

**Comment by DhairyaBahl** [comment]
@morevnaproject @Gilraiser I understand that, I have to create and add the feature which is present in the 3rd gif. but to do that can you guys guide me a bit that what that new feature is doing ?? how can we do the same thing now without that feature so that I will understand that what actions are going on when we click that key icon ?? 

Thanks :)

PS: @Keyikedalube Can you help me a bit in this area just a bit of where to refer and where to start it. kind of thing ?? Thanks  :)

**Comment by Keyikedalube** [comment]
@DhairyaBahl If I had to start working with this issue assuming not knowing where to start.
I'd do the following... first, see:
1. where the onion toggle button source code is
2. check its toggle button clicked signal
3. and voila! You have the internal logic already in reach

> what that new feature is doing ??

Same behavior as [onion skin](https://en.wikipedia.org/wiki/Onion_skinning). In Synfig, we have **future** and **past** frames (second spinbutton not seen in the gif above) feature. However, this time the user wants this feature implemented for keyframes.

Look at the first gif above. Notice three black inverted triangles in the timeline? The user wants to see only those frames set as keyframe and not the entire frames like onion skin does.

In simple words, onion skin fetches frame-by-frame. Open synfigstudio, place your timeline cursor at 1s, and hit comma `,` key once. That's one frame. If you press that key again, that's second frame. So, that's frame by frame.

This time press left bracket `[` key and your cursor will be taken directly to the first keyframe.
So, depending on how many keyframes a user has, toggle keyframe feature will fetch keyframe-by-keyframe.

**Comment by DhairyaBahl** [comment]
@keyikedalube Thanks you for giving the guidance and to explain every term in brief :) . Now I will be able to move forward in solving the issue.

**Comment by DhairyaBahl** [comment]
@Keyikedalube  

* what I understood
I am able to understand the use of onion skin. It shows the "just previous" frame along with the current frame in a lighter version and I am able to locate its source code inn the canvasview.cpp and its action is being defined in the app.cpp. Now I am trying to figure out how the whole feature is actually working.

* what i wanted to ask 
1. Do we just need to create an icon and by toggling it, onion skill will now work as keyframe onion skin rather than normal one ?? Will that work ?? 

2.  What exactly is happening in the first gif ?? We first separated the two keyframes onion skin and then we created a new figure by somehow joining them ? 

Side Note: This is my biggest issue till now, so kindly don't mind if I ask more questions .

Thank You :)

**Comment by rodolforg** [comment]
A possible tip:
As far as I recall to read somewhere (here or forums): the onion skin was by frame and 'recently' changed to by keyframe.
If you are able to find that in a git log of a (related) modified file, it should be easy to know how to change between those onion-skin styles.

**Comment by DhairyaBahl** [comment]
@rodolforg I will definitely look into previous commits by git log . Thanks for the information !! :)

**Comment by Keyikedalube** [comment]
> 1. Do we just need to create an icon and by toggling it, onion skill will now work as keyframe onion skin rather than normal one ?? Will that work ??

Yes :)
But you have to implement the signal clicked logic to keyframes just like onion skin does for frames
Code could be different from what I'm suggesting above. Might take a look at it later when I've free time

> 2. What exactly is happening in the first gif ?? We first separated the two keyframes of onion skin and then we created a new figure by somehow joining them ?

The answer to your *then we created a new figure by somehow joining them* is No
Here's what the OP did for each keyframe
1. Circle
2. Distorted circle
3. Oval

He's animating a circle changing to an oval shape. Each of those three keyframes is done manually by him. Not joined and Synfig cannot create that unique **distorted circle** shape by itself LOL. The user has to draw it (specifically change the shape from **circle**) first.

**Comment by Keyikedalube** [comment]
> Code could be different from what I'm suggesting above. Might take a look at it later when I've free time

@DhairyaBahl take note of spinbuttons too. I checked out spinbuttons `->signal_value_changed()` code

- The signal. **Get the no of keyframes**
https://github.com/synfig/synfig/blob/1b9ec56dd3234ae403cacc908a5692e45f4b1497/synfig-studio/src/gui/canvasview.cpp#L2636-L2645
- Further down. **Send the data to synfigapp api**
https://github.com/synfig/synfig/blob/1b9ec56dd3234ae403cacc908a5692e45f4b1497/synfig-studio/src/gui/workarea.cpp#L768-L775
- Final destination
`save_meta_data()` I believe this is where you're going to dig deeper/study the internals. Hope I'm pointing in the right direction @ice0 @rodolforg?

**Comment by DhairyaBahl** [comment]
@Keyikedalube Thank You so much for helping. I will look deeper in these files !! Thanks for direction :)

**Comment by DhairyaBahl** [comment]
@ice0 @Keyikedalube @rodolforg 
I am getting doubts at every step. What should I do when I stuck ??

Like in this 
```
void
CanvasView::set_onion_skins()
{
	if(toggling_onion_skin)                  <-    <-    <-    <-    <-    <-    <-    <- this line 
		return;
	int onion_skins[2];
	onion_skins[0]=past_onion_spin->get_value();
	onion_skins[1]=future_onion_spin->get_value();
	work_area->set_onion_skins(onion_skins);
}
```
where should I look for the ```get_value function ``` and the condition in the specified line should be that if toggling_onion_skin is false then it should then return otherwise it should continue with the function but it is the opposite !! I know i am wrong but I am unable to understand this part ??

Also I have researched other files but I am unable to get it where we have specified that onion_skin spinbutton will work frame-by-frame not by keyframe by keyframe !?

Kindly help me to move further.

Thanks :)

**Comment by rodolforg** [comment]
I guess this line is about interlocking due to some odd/ugly setup or to avoid some recursive or excessive calls while user sets some onion skin values.
With git grep (or with a search for this word or symbol references in a IDE) you can see when this variable is set to true and false.

Normally this could be unnecessary if the `set_xxxx()` method first checks if the values really changed before do any actual action.
----
I'm a little bit lazy to read all discussion by now, but I'll try to help you too. There's two things to do here:

1. Provide a button to toggle between onion skinning by regular frames or by keyframes
2. Provide support for both cases of onion skinning rendering.

For the first item, you should be a little used to gtkmm. If you're not used to it, I suggest you to read https://developer.gnome.org/gtkmm-tutorial/stable/, specially sections Basics, Buttons and Appendix Signals (B1).
This button would essentially change/toggle a new variable of WorkArea to store what onion skinning method user wants.

The second item is about the real action and the code is elsewhere: in Renderer_Canvas. There, you can see when they search computes all frames the will be rendered, and change how they're chosen.

A third step would be adding this user choice either in user preferences or synfig file metadata, I don't know which one.

**Comment by Gilraiser** [comment]
Great, thank you so much

**Comment by morevnaproject** [comment]
Resolved in Synfig version 1.5.1 - https://www.synfig.org/2021/10/30/synfig-studio-1-5-1-released/
